addons/mail/models/mail_link_preview.py

```diff
@@ -0,0 +1,137 @@
+    image_url = fields.Char('image_url')
+    description = fields.Text('description')
+    mimetype = fields.Char('mimetype')
+
+    def _link_preview_format(self):
+        return [{
+            'description': preview.description,
+            'id': preview.id,
+            'image_url': preview.image_url,
+            'mimetype': preview.mimetype,
+            'title': preview.title,
+            'type': preview.type,
+            'url': preview.url,
+        } for preview in self]
+
+    def _throttle_link_preview(self, url):
+        throttle = self.env['ir.config_parameter'].sudo().get_param('mail.link_preview_throttle', 99)
+        if int(throttle) == 0:
+            return True
+        domain = urlparse(url).netloc
+        date_interval = fields.Datetime.to_string((datetime.now() - relativedelta(seconds=10)))
+        call = self.search_count([
+            ('url', 'like', '%' + domain + '%'),
+            ('write_date', '>', date_interval),
+        ])
+        return call > int(throttle)
+
+    @api.model
+    def _create_link_preview(self, url):
+        if self._throttle_link_preview(url):
+            return False
+        link_preview_data = self._get_link_preview_from_url(url)
+
+        if link_preview_data:
+            link_preview = self.sudo().create(link_preview_data)
+            return [('insert', link_preview._link_preview_format()[0])]
+        return False
+
+    @api.model
+    def _get_link_preview_from_url(self, url):
+        try:
+            response = requests.head(url, timeout=3)
+        except requests.exceptions.RequestException:
+            return False
+
+        if response.status_code != requests.codes.ok:
+            return False
+
+        image_mimetype = [
+            'image/bmp',
+            'image/gif',
+            'image/jpeg',
+            'image/png',
+            'image/tiff',
+            'image/x-icon',
+        ]
+        # Content-Type header can return a charset, but we just need the
+        # mimetype (eg: image/jpeg;charset=ISO-8859-1)
+        content_type = response.headers['Content-Type'].split(';')
+        if content_type[0] in image_mimetype:
+            return {
+                'description': False,
+                'image_url': url,
+                'mimetype': content_type[0],
+                'title': url,
+                'type': 'image',
+                'url': url,
+            }
+        if 'text/html' in response.headers['Content-Type']:
+            return self._get_link_preview_from_html(url)
+        return False
+
+    def _get_link_preview_from_html(self, url):
+        headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; rv:91.0) Gecko/20100101 Firefox/91.0'}
+        response = requests.get(url, timeout=3, stream=True, headers=headers)
+        parser = etree.HTMLParser(encoding=response.encoding)
+        tree = html.fromstring(response.content, parser=parser)
+        title = tree.xpath('//meta[@property="og:title"]/@content')
+        if title:
```

Comment:
My logic was that we might not want to create a link preview that has no title since it's one of the 4 required properties.
(see https://ogp.me/#metadata).
This is not really important I guess.

Reply:
But then there is no reason to treat title more particularly than the other required properties, so current condition still feels strange.